Add Kubernetes API server SLOs (#12)
appuio/component-openshift4-slos: docs/modules/ROOT/pages/runbooks/kubernetes_api.adoc

= Kubernetes API SLOs

include::partial$runbooks/contribution_note.adoc[]

[[requests]]
== API Requests

=== icon:glasses[] Overview

This SLO measures the percentage of valid, well structured Kubernetes API requets that fail.
The error rate is a general indicator of the health of the API server, but might not show you the root cause of the issue.

include::partial$runbooks/alert_types.adoc[]

=== icon:bug[] Steps for debugging

A high Kubernetes API server error rate can have multiple root causes.
First check if you are generally able to connect the cluter's API through `kubectl`.

> If you can't, you most likely also received a `KubeApiServerFailure` and it probably makes sense to also look at its runbook even if you didn't.

Reviewer: ```suggestion
If you can't, you most likely also received a `KubeApiServerFailure` and it makes sense to also look at its runbook even if the alert did not trigger.
```

If you can still access the API server, look at the pods running in namespace `openshift-kube-apiserver`.

> Check if any of the API server pods seem to be crashing and ceck their logs.

Reviewer: ```suggestion
Check if any of the API server pods seem to be crashing and check their logs.
```

Also check etcd running in namespace `openshift-etcd` and see if any pods are crashing or logging errors.

NOTE: If you have issues with viewing logs through `kubectl`, but the cluster generally still works, you might be able to look at the logs in elasticsearch.

The API server logs can be fairly noisy, to help you narrow down the issue, you can connect to the Prometheus on the cluster and run the following query:

[source,promql]
----
sum (rate(apiserver_request_total{code=~"(5..|429)"}[10m])) by (verb, resource, code)
----

This should return one or more time series.
With these you should be able to narrow down the issue:

* If all, or most, timeseries have the code `429` this most likely means the API server is overloaded.
In that case, double-check if one or more master nodes have a high load.
If so either investigate what generates the high load or increase master nodes size.

> * If you have a large amount of `504` errors, an upstream service is miss-behaving.

Reviewer: ```suggestion
* If you have a large amount of `504` errors, an upstream service is misbehaving.
```

If they're running, but the Kubernetes API is still not reachable, try to connect to one of them using SSH.
You'll need the SSH key stored in vault and use one of the LB VMs as a jumphost.

[source,shell]
----
# For example: https://api.syn.vshn.net
# IMPORTANT: do NOT add a trailing `/`. Commands below will fail.
export COMMODORE_API_URL=<lieutenant-api-endpoint>

# Set Project Syn cluster and tenant ID
export CLUSTER_ID=<lieutenant-cluster-id> # Looks like: c-<something>
export TENANT_ID=$(curl -sH "Authorization: Bearer $(commodore fetch-token)" ${COMMODORE_API_URL}/clusters/${CLUSTER_ID} | jq -r .tenant)

# Login to vault
export VAULT_ADDR=https://vault-prod.syn.vshn.net
vault login -method=oidc

# Fetch SSH key
vault kv get -format=json clusters/kv/${TENANT_ID}/${CLUSTER_ID}/cloudscale/ssh \
  | jq -r '.data.data.private_key' | base64 --decode > ssh_key
chmod 400 ssh_key

# Connect to master node
MASTER_NODE=etcd-0
LB_HOST=$(grep -E "^Host.*${CLUSTER_ID}" ~/.ssh/sshop_config | head -1 | awk '{print $2}')
ssh -J "${LB_HOST}" -i ssh_key "core@${MASTER_NODE}"
----

Check the logs in `/var/log/etcd`, `/var/log/kube-apiserver`, and `/var/log/containers`.

> Also see if any systemd service as crashed.

Reviewer: ```suggestion
Also see if any systemd service has crashed.
```
